components/chainhook-postgres/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+pub mod types;
+pub mod utils;
+
+use std::future::Future;
+
+pub use deadpool_postgres;
+use deadpool_postgres::{Manager, ManagerConfig, Object, Pool, RecyclingMethod, Transaction};
+pub use tokio_postgres;
+
+use tokio_postgres::{Client, Config, NoTls, Row};
+
+/// A Postgres configuration for a single database.
+#[derive(Clone, Debug)]
+pub struct PgConnectionConfig {
+    pub dbname: String,
+    pub host: String,
+    pub port: u16,
+    pub user: String,
+    pub password: Option<String>,
+    pub search_path: Option<String>,
+}
+
+/// Creates a Postgres connection pool based on a single database config. You can then use this pool to create ad-hoc clients and
+/// transactions for interacting with the database.
+pub fn new_pg_connection_pool(config: &PgConnectionConfig) -> Result<Pool, String> {
+    let mut pg_config = Config::new();
+    pg_config
+        .dbname(&config.dbname)
+        .host(&config.host)
+        .port(config.port)
+        .user(&config.user)
+        .options(format!(
+            "-csearch_path={}",
+            config.search_path.as_ref().unwrap_or(&"public".to_string())
+        ));
+    if let Some(password) = &config.password {
+        pg_config.password(password);
+    }
+    let manager = Manager::from_config(
+        pg_config,
+        NoTls,
+        ManagerConfig {
+            recycling_method: RecyclingMethod::Fast,
+        },
+    );
+    Ok(Pool::builder(manager)
+        .max_size(16)
+        .build()
+        .map_err(|e| format!("unable to build pg connection pool: {e}"))?)
+}
+
+/// Creates a short-lived client and passes it into the given closure. Takes the connection out of the given pool, so it may or
+/// may not close once the given closure is complete.
+pub async fn with_pg_client<F, Fut, T>(pool: &Pool, f: F) -> Result<T, String>
+where
+    F: FnOnce(Object) -> Fut,
+    Fut: Future<Output = Result<T, String>>,
+{
+    let client = pool
+        .get()
+        .await
+        .map_err(|e| format!("unable to get pg client: {e}"))?;
+    let result = f(client).await?;
+    Ok(result)
+}
+
+/// Takes a connection from the given pool, opens a transaction, and passes it into the given closure. Use this when you want to
+/// run queries within the confines of a single well defined transaction.
+pub async fn with_pg_transaction<'a, F, Fut, T>(pool: &Pool, f: F) -> Result<T, String>
+where
+    F: FnOnce(&'a Transaction<'a>) -> Fut,
+    Fut: Future<Output = Result<T, String>> + 'a,
+{
+    let mut client = pool
+        .get()
+        .await
+        .map_err(|e| format!("unable to get pg client: {e}"))?;
+    let transaction = client
+        .transaction()
+        .await
+        .map_err(|e| format!("unable to begin pg transaction: {e}"))?;
+    {
+        let transaction_ref: &'a Transaction<'a> = unsafe { std::mem::transmute(&transaction) };
```

Comment:
This looks spooky, I'll just assume it's all good

Reply:
Oh, good catch, I remember doing that to get over some compiler thing. I'll take another look to see if there's a better way to handle the `Transaction` lifetime